AppController/djinn.rb

```diff
@@ -619,19 +619,36 @@ def get_app_info_map(secret)
   #   "OK" if the relocation occurred successfully, and a String containing the
   #   reason why the relocation failed in all other cases.
   def relocate_app(appid, http_port, https_port, secret)
-    if !valid_secret?(secret)
-      return BAD_SECRET_MSG
+    return BAD_SECRET_MSG if !valid_secret?(secret)
+    Djinn.log_debug("Received relocate_app for #{appid} for " +
+        " http port #{http_port} and https port #{https_port}.")
+
+    if !my_node.is_shadow?
+      # We need to send the call to the shadow.
+      Djinn.log_debug("Sending relocate_app call for #{appid} to shadow.")
+      acc = AppControllerClient.new(get_shadow.private_ip, @@secret)
+      begin
+        return acc.relocate_app(appid, http_port, https_port)
+      rescue FailedNodeException => except
+        Djinn.log_warn("Failed to forward relocate_app call to shadow (#{get_shadow}).")
+        return NOT_READY
+      end
     end
 
-    # Only the login node runs relocate.
-    if not my_node.is_login?
-      Djinn.log_warn("Only login nodes runs relocate.")
-      return "Error: this is not the login node."
+    # Some sanity check on parameter and app.
+    begin
+      http_port = Integer(http_port)
+      https_port = Integer(https_port)
+    rescue ArgumentError
+      Djinn.log_warn("relocate_app received invalid port values.")
+      return INVALID_REQUEST
+    end
+    if @app_info_map[appid].nil? or @app_info_map[appid]['nginx'].nil? or
+        @app_info_map[appid]['nginx_https'].nil? or
+        @app_info_map[appid]['haproxy'].nil?
+      Djinn.log_warn("Received relocate request for non running app #{appid}.")
```

Comment:
This message might be misleading. I suggest you change it you something like:

"Unable to relocate due to missing app settings for: #{appid}"